README.md
# Laravel Sniptbot

This repository implements a simple [ServiceProvider](https://laravel.com/docs/master/providers) that makes the instances of the Snippet **Crawler** and **Sniffer** easily accessible via a [Facade](https://laravel.com/docs/master/facades) in [Laravel](http://laravel.com). See [@Snippetify/Snippet-sniffer](https://github.com/snippetify/snippet-sniffer) for more information about the PHP Snippet sniffing and its interfaces.

## What it does

This library allows you 

1. To get code snippets using search engine api (Google)
2. To get code snippets from any web page by crawling url seeds.

## How to use it

```bash
$ composer require snippetify/laravel-sniptbot
```

```php
php artisan vendor:publish --provider="Snippetify\Sniptbot\SniptbotServiceProvider"
```

##### Configuration

Open the created file in the `config/sniptbot.php` and customize the configuration options to your liking.

```php
return [
    // Search engine api configuration keys
  	'provider' 	=> [
      // Your google Search engine ID
      // https://developers.google.com/custom-search/v1/introduction
    	'cx' 	=> env('SNIPTBOT_PROVIDER_CX', ''),
      // Your google API key
      // https://developers.google.com/custom-search/v1/overview#api_key
    	'key' 	=> env('SNIPTBOT_PROVIDER_KEY', ''),
    	'name' 	=> env('SNIPTBOT_PROVIDER_NAME', 'google'), // provider name (google)
  	],
  	
  	// Optional
  	// Useful for adding meta information to each snippet
  	'app' => [
    	'name' 		=> env('SNIPTBOT_NAME', 'Sniptbot'), // your sniptbot name
    	'version' 	=> env('SNIPTBOT_VERSION', '1.0.0'), // your sniptbot version
  	],

  	// Optional
  	// Useful for logging
  	'logger' => [
    	'name' => env('SNIPTBOT_NAME', 'Sniptbot'), // logger name
    	'file' => storage_path('logs/sniptbot.log'), // logger file path
  	],

  	// Optional
  	// Useful for scraping
  	'html_tags' => [
    	'index' 	=> 'h1, h2, h3, h4, h5, h6, p, li', // Tags to index
    	'snippet' 	=> 'pre[class] code, div[class] code, .highlight pre, code[class]', // Tags to fetch snippets
  	],

  	// Optional
  	// Useful for adding new scrapers
  	// The name must be the website host without the scheme i.e. not https://foo.com but foo.com
  	'scrapers' => [
    	// 'scraper_name' 	 => ScraperClass::class, // You can add as many as you want
  	],

  	// Optional
  	// Useful for adding new providers
  	'providers' => [
    	// 'provider_name' 	=> ProviderClass::class, // You can add as many as you want
  	],

  	// Optional
  	// Useful for web crawling
  	// Please follow the link below for more information as we use Spatie crawler
  	// https://github.com/spatie/crawler
  	'crawler' => [
    	// 'langs' 				 => ['en'],
    	// 'profile' 				 => Snippetify\SnippetSniffer\Profiles\CrawlSubdomainsAndUniqueUri::class,
    	// 'user_agent' 			 => Snippetify\SnippetSniffer\Core::CRAWLER_USER_AGENT,
    	// 'concurrency' 			 => 10,
    	// 'ignore_robots' 		 => false,
    	// 'maximum_depth' 		 => null,
    	// 'execute_javascript' 	 => false,
    	// 'maximum_crawl_count' 	 => null,
    	// 'parseable_mime_types' 	 => ['text/html'],
    	// 'maximum_response_size'  => 1024 * 1024 * 3,
    	// 'delay_between_requests' => 250,
  	]
];
```

#### Sniffing from search engine api

```php
use Snippetify\Sniptbot\Facades\Sniffer;

// Important: You must set your search engine keys before using it
// @return Snippetify\SnippetSniffer\Common\Snippet[]
$snippets = Sniffer::fetch('js array contains value');
```

#### Crawl url seeds

```php
use Snippetify\Sniptbot\Facades\Crawler;

// @return Snippetify\SnippetSniffer\Common\MetaSnippetCollection[]
$snippets = Crawler::fetch(['your urls']);
```



## Changelog

Please see [CHANGELOG](https://github.com/snippetify/laravel-sniptbot/blob/master/CHANGELOG.md) for more information what has changed recently.

## Testing

 You must set the **PROVIDER_NAME**, **PROVIDER_CX**, **PROVIDER_KEY**, **CRAWLER_URI**, keys in phpunit.xml file before running tests.

**Important:** Those links must contains at least one snippet otherwise the tests will failed.

```bash
composer test
```

## Contributing

Please see [CONTRIBUTING](https://github.com/snippetify/laravel-sniptbot/blob/master/CONTRIBUTING.md) for details.

## Credits

1. [Evens Pierre](https://github.com/pierrevensy)

## License

The MIT License (MIT). Please see [License File](https://github.com/snippetify/laravel-sniptbot/blob/master/LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All notable changes to **[sniptbot](https://github.com/snippetify/laravel-sniptbot)** will be documented in this file

## 1.0.0 - 2020-06-30

- initial release



tests/TestCase.php
<?php

/*
 * This file is part of the snippetify package.
 *
 * (c) Evens Pierre <evenspierre@snippetify.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Snippetify\Sniptbot\Tests;

use Snippetify\Sniptbot\Facades\Crawler;
use Snippetify\Sniptbot\Facades\Sniffer;
use Snippetify\Sniptbot\SniptbotServiceProvider;

class TestCase extends \Orchestra\Testbench\TestCase
{
    protected function getPackageProviders($app)
    {
        return [
            SniptbotServiceProvider::class
        ];
    }

    protected function getPackageAliases($app)
    {
        return [
            'Crawler' => Crawler::class,
            'Sniffer' => Sniffer::class
        ];
    }
}


tests/SniptbotServiceProviderTest.php
<?php

/*
 * This file is part of the snippetify package.
 *
 * (c) Evens Pierre <evenspierre@snippetify.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Snippetify\Sniptbot\Tests;

use Snippetify\SnippetSniffer\WebCrawler;
use Snippetify\SnippetSniffer\SnippetSniffer;

class SniptbotServiceProviderTest extends TestCase
{
    /**
     * @test
     */
    public function has_registered_services()
    {
        $this->assertTrue($this->app->bound('sniptbot.crawler'));
        $this->assertTrue($this->app->bound('sniptbot.sniffer'));

        $this->assertInstanceOf(WebCrawler::class, $this->app->make('sniptbot.crawler'));
        $this->assertInstanceOf(SnippetSniffer::class, $this->app->make('sniptbot.sniffer'));
    }

    /**
     * @test
     */
    public function has_registered_aliases()
    {
        $this->assertTrue($this->app->isAlias(WebCrawler::class));
        $this->assertTrue($this->app->isAlias(SnippetSniffer::class));
        $this->assertEquals('sniptbot.crawler', $this->app->getAlias(WebCrawler::class));
        $this->assertEquals('sniptbot.sniffer', $this->app->getAlias(SnippetSniffer::class));
    }

    /**
     * @test
     */
    public function has_registered_package_config()
    {
        $config = $this->app->make('config');

        $this->assertEquals($config->get('sniptbot.app'), [
            'name' => 'Sniptbot',
            'version' => '1.0.0',
        ]);
    }

    /**
     * @test
     */
    public function does_not_provide_singleton_instance()
    {
        $this->assertNotSame($this->app->make('sniptbot.crawler'), $this->app->make('sniptbot.crawler'));
        $this->assertNotSame($this->app->make('sniptbot.sniffer'), $this->app->make('sniptbot.sniffer'));
    }

    /**
     * @test
     */
    public function does_sniffer_contains_results()
    {
        $data = $this->app->make('sniptbot.sniffer')->fetch('js array contains', [ 'page' => 1, 'limit' => 10 ]);

        $this->assertGreaterThan(0, count($data));
    }

    /**
     * @test
     */
    public function does_crawler_contains_results()
    {
        $data = $this->app->make('sniptbot.crawler')->fetch([$_SERVER['CRAWLER_URI']]);

        $this->assertGreaterThan(0, count($data));
    }
}


src/SniptbotServiceProvider.php
<?php
/*
 * This file is part of the snippetify package.
 *
 * (c) Evens Pierre <evenspierre@snippetify.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Snippetify\Sniptbot;

use Illuminate\Support\ServiceProvider;
use Snippetify\SnippetSniffer\WebCrawler;
use Snippetify\SnippetSniffer\SnippetSniffer;
use Illuminate\Contracts\Support\DeferrableProvider;

class SniptbotServiceProvider extends ServiceProvider implements DeferrableProvider
{
    /**
     * Get the path of the configuration file shipping with the package.
     *
     * @return string
     */
    public function getConfigPath()
    {
        return dirname(__DIR__) . '/config/sniptbot.php';
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        $this->publishConfig($this->getConfigPath());
    }

    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        $this->registerConfig();
        $this->registerServices();
        $this->registerAliases();
    }

    /**
     * Register a path to be published by the publish command.
     *
     * @param string $path
     * @param string $group
     * @return void
     */
    protected function publishConfig($path, $group = 'config')
    {
        $this->publishes([$path => config_path('sniptbot.php')], $group);
    }

    /**
     * Register the default configuration.
     *
     * @return void
     */
    protected function registerConfig()
    {
        $this->mergeConfigFrom($this->getConfigPath(), 'sniptbot');
    }

    /**
     * Register the Goutte instance.
     *
     * @return void
     */
    protected function registerServices()
    {
        $this->app->bind('sniptbot.crawler', function ($app) {
            $config = $app->make('config');
            return new WebCrawler($config->get('sniptbot', []));
        });

        $this->app->bind('sniptbot.sniffer', function ($app) {
            $config = $app->make('config');
            return new SnippetSniffer($config->get('sniptbot', []));
        });
    }

    /**
     * Register class aliases.
     *
     * @return void
     */
    protected function registerAliases()
    {
        $this->app->alias('sniptbot.crawler', WebCrawler::class);
        $this->app->alias('sniptbot.sniffer', SnippetSniffer::class);
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return [
            'sniptbot.crawler',
            'sniptbot.sniffer',
        ];
    }
}


config/sniptbot.php
<?php

/*
 * This file is part of the snippetify package.
 *
 * (c) Evens Pierre <evenspierre@snippetify.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

return [

    /*
     |-----------------------------------------------------------------------
     | Snippetify Sniptbot Configuration
     |-----------------------------------------------------------------------
     |
     | A collection of default options to apply to snippet sniffer objects
     |
     | @see {@link https://github.com/snippetify/snippet-sniffer}
     */

    // Search engine api configuration keys
  	'provider' 	=> [
      // Your google Search engine ID
      // https://developers.google.com/custom-search/v1/introduction
    	'cx' 	=> env('SNIPTBOT_PROVIDER_CX', ''),
      // Your google API key
      // https://developers.google.com/custom-search/v1/overview#api_key
    	'key' 	=> env('SNIPTBOT_PROVIDER_KEY', ''),
    	'name' 	=> env('SNIPTBOT_PROVIDER_NAME', 'google'), // provider name (google)
  	],
  	
  	// Optional
  	// Useful for adding meta information to each snippet
  	'app' => [
    	'name' 		=> env('SNIPTBOT_NAME', 'Sniptbot'), // your sniptbot name
    	'version' 	=> env('SNIPTBOT_VERSION', '1.0.0'), // your sniptbot version
  	],

  	// Optional
  	// Useful for logging
  	'logger' => [
    	'name' => env('SNIPTBOT_NAME', 'Sniptbot'), // logger name
    	'file' => storage_path('logs/sniptbot.log'), // logger file path
  	],

  	// Optional
  	// Useful for scraping
  	'html_tags' => [
    	'index' 	=> 'h1, h2, h3, h4, h5, h6, p, li', // Tags to index
    	'snippet' 	=> 'pre[class] code, div[class] code, .highlight pre, code[class]', // Tags to fetch snippets
  	],

  	// Optional
  	// Useful for adding new scrapers
  	// The name must be the website host without the scheme i.e. not https://foo.com but foo.com
  	'scrapers' => [
    	// 'scraper_name' 	 => ScraperClass::class, // You can add as many as you want
  	],

  	// Optional
  	// Useful for adding new providers
  	'providers' => [
    	// 'provider_name' 	=> ProviderClass::class, // You can add as many as you want
  	],

  	// Optional
  	// Useful for web crawling
  	// Please follow the link below for more information as we use Spatie crawler
  	// https://github.com/spatie/crawler
  	'crawler' => [
    	// 'langs' 				 => ['en'],
    	// 'profile' 				 => Snippetify\SnippetSniffer\Profiles\CrawlSubdomainsAndUniqueUri::class,
    	// 'user_agent' 			 => Snippetify\SnippetSniffer\Core::CRAWLER_USER_AGENT,
    	// 'concurrency' 			 => 10,
    	// 'ignore_robots' 		 => false,
    	// 'maximum_depth' 		 => null,
    	// 'execute_javascript' 	 => false,
    	// 'maximum_crawl_count' 	 => null,
    	// 'parseable_mime_types' 	 => ['text/html'],
    	// 'maximum_response_size'  => 1024 * 1024 * 3,
    	// 'delay_between_requests' => 250,
  	]
];
